service/history/api/respondactivitytaskcompleted/api.go

```diff
@@ -97,13 +101,25 @@ func Invoke(
 			}
 
 			if !isRunning ||
-				ai.StartedEventId == common.EmptyEventID ||
+				(!isCalledCompleteActivityById && ai.StartedEventId == common.EmptyEventID) ||
 				(token.GetScheduledEventId() != common.EmptyEventID && token.Attempt != ai.Attempt) ||
 				(token.GetVersion() != common.EmptyVersion && token.Version != ai.Version) {
 				return nil, consts.ErrActivityTaskNotFound
 			}
 
-			if _, err := mutableState.AddActivityTaskCompletedEvent(scheduledEventID, ai.StartedEventId, request); err != nil {
+			// We fabricate a started event only when the activity is not started yet and
+			// we need to force complete an activity
+			if isCalledCompleteActivityById && ai.GetRetryLastFailure() != nil {
```

Comment:
This condition is irrelevant `ai.GetRetryLastFailure() != nil`.

Replace with `ai.StartedEventId == common.EmptyEventID`.

Reply:
A question here, if an activity is not started yet, the `ai.StartedEventId` will always be empty?

The reason I asked about this is because if there is no start event then we will fabricate a started event here to match the complete event.

Reply:
> A question here, if an activity is not started yet, the ai.StartedEventId will always be empty?

That is my understanding, yes.